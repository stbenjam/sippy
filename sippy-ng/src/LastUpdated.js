import PropTypes from 'prop-types'
import React, { Fragment } from 'react'

export default function LastUpdated (props) {
  const getLastUpdated = () => {
    const rtf = new Intl.RelativeTimeFormat('en', { numeric: 'auto' })
    const millisAgo = props.lastUpdated ? props.lastUpdated.getTime() - Date.now() : 0

    const minute = 1000 * 60 // Milliseconds in a minute
    const hour = 60 * minute // Milliseconds in an hour

    if (millisAgo === 0) {
      return <Fragment>unknown</Fragment>
    } else if (Math.abs(millisAgo) < hour) {
      return (
                <Fragment>
                    {rtf.format(Math.round(millisAgo / minute), 'minutes')}
                </Fragment>
      )
    } else {
      return (
                <Fragment>
                    {rtf.format(Math.round(millisAgo / hour), 'hours')}
                </Fragment>
      )
    }
  }

  return <Fragment>Last updated {getLastUpdated()}</Fragment>
}

LastUpdated.propTypes = {
  lastUpdated: PropTypes.instanceOf(Date).isRequired
}
