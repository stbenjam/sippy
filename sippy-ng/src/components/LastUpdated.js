import PropTypes from 'prop-types'
import React, { Fragment } from 'react'

export default function LastUpdated (props) {
  const rtf = new Intl.RelativeTimeFormat('en', { numeric: 'auto' })
  const millisAgo = props.lastUpdated ? (props.lastUpdated.getTime() - Date.now()) : 0

  const minute = 1000 * 60 // Milliseconds in a minute
  const hour = 60 * minute // Milliseconds in an hour

  try {
    if (millisAgo === 0) {
      return <Fragment>unknown</Fragment>
    } else if (Math.abs(millisAgo) < hour) {
      return (
                <Fragment>
                    Last updated {rtf.format(Math.round(millisAgo / minute), 'minutes')}
                </Fragment>
      )
    } else {
      return (
                <Fragment>
                    Last updated {rtf.format(Math.round(millisAgo / hour), 'hours')}
                </Fragment>
      )
    }
  } catch (e) {
    return <></>
  }
}

LastUpdated.propTypes = {
  lastUpdated: PropTypes.instanceOf(Date).isRequired
}
