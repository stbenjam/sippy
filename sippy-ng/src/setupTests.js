/** @jest-environment setup-polly-jest/jest-environment-node */

import '@testing-library/jest-dom'
import Enzyme from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import { Polly } from '@pollyjs/core'
import HttpAdapter from '@pollyjs/adapter-node-http'
import FSPersister from '@pollyjs/persister-fs'
import fetch from 'node-fetch'

// https://github.com/mui-org/material-ui/issues/21293
export const withoutMuiID = (wrapper) => wrapper.html().replace(/id="mui-[0-9]*"/g, '').replace(/aria-labelledby="(mui-[0-9]* *)*"/g, '')

Enzyme.configure({ adapter: new Adapter() })

global.fetch = fetch

// See: https://github.com/vuejs/vue-test-utils/issues/974
global.requestAnimationFrame = cb => cb()
