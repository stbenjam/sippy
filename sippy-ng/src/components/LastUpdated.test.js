import React from 'react'
import { act } from '@testing-library/react'
import toJson from 'enzyme-to-json'
import { withoutMuiID } from '../setupTests'
import LastUpdated from './LastUpdated'
import { mount } from 'enzyme'

const minute = 60 * 1000
const hour = minute * 60

jest.useRealTimers()

describe(LastUpdated, () => {
  it('shows minutes when < an hour', async () => {
    const lastUpdated = new Date(0)
    Date.now = jest.fn(() => 10 * minute)

    let wrapper
    await act(async () => {
      wrapper = mount(<LastUpdated lastUpdated={lastUpdated} />)
    })

    expect(wrapper.html()).toContain('Last updated 10 minutes ago')
    expect(toJson(withoutMuiID(wrapper))).toMatchSnapshot()
  })

  it('shows hours when > an hour', async () => {
    const lastUpdated = new Date(0)
    Date.now = jest.fn(() => 3 * hour)

    let wrapper
    await act(async () => {
      wrapper = mount(<LastUpdated lastUpdated={lastUpdated} />)
    })

    expect(wrapper.html()).toContain('Last updated 3 hours ago')
    expect(toJson(withoutMuiID(wrapper))).toMatchSnapshot()
  })
})
