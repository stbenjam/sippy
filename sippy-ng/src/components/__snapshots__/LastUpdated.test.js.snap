// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastUpdated shows hours when > an hour 1`] = `undefined`;

exports[`LastUpdated shows minutes when < an hour 1`] = `undefined`;
